### PR TITLE
[SPARK-24297][CORE] Fetch-to-disk by default for > 2gb

### PR DESCRIPTION
**Pull request by squito**
Fetch-to-mem is guaranteed to fail if the message is bigger than 2 GB,
so we might as well use fetch-to-disk in that case.  The message includes
some metadata in addition to the block data itself (in particular
UploadBlock has a lot of metadata), so we leave a little room.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (squito) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: no guarantee its big enough.  Seemed OK in the test I tried.  But UploadBlock has some variable length strings so can't say for sure.

I'm fine making this much bigger, eg. 1 MB -- you'd only be bigger than that with a pathological case.  then there would be *some* cases where we'd be taking an old message which was fine with fetch-to-mem and we'd switch to fetch-to-disk.  But such a tiny case, and not an unreasonable change even for that ... so should be OK.

**Review by felixcheung — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Actually I prefer 512 to 500  :)

Comment (squito) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: sounds good, updated

**Comment by HyukjinKwon** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by jiangxb1987** [comment]
LGTM

**Comment by HyukjinKwon** [comment]
@jerryshao, busy? I think you better take a look before merging this in if you are available.

**Comment by jerryshao** [comment]
I will take a look at this sometime day, but don't block on me if it is urgent.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: I think the original purpose to set to `Long.MaxValue` is to avoid using this configuration by default, user should set to a proper size to enable this feature. But anyway I think the current change is also fine.

**Comment by HyukjinKwon** [comment]
Still LGTM but let me leave it to you @squito and @jerryshao .

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please
--

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by jerryshao** [comment]
Hi @squito , would you please also update the changes in the doc, thanks!

**Comment by squito** [comment]
good point @jerryshao , I've updated the docs now, please take a look, thanks

**Review by jerryshao — APPROVED** [review]
LGTM, merging to master.

**Comment by tgravescs** [comment]
just to clarify here, so the failure case if you are using a shuffle server < 2.2, does it throw an obvious error to the user?

**Comment by squito** [comment]
oh no, good point Tom.  it'll fail with

```
18/07/26 07:15:02 WARN scheduler.TaskSetManager: Lost task 1.0 in stage 1.3 (TID 15, irashid-large-2.gce.cloudera.com, executor 2): FetchFailed(BlockManagerId(1
, irashid-large-3.gce.cloudera.com, 7337, None), shuffleId=0, mapId=1, reduceId=1, message=
org.apache.spark.shuffle.FetchFailedException: java.lang.UnsupportedOperationException
        at org.apache.spark.network.server.StreamManager.openStream(StreamManager.java:60)
        at org.apache.spark.network.server.TransportRequestHandler.processStreamRequest(TransportRequestHandler.java:136)
...
```

lemme open a jira to improve this.

**Pull request closed without merging**
